win32ss/user/user32/controls/ghost.c

```diff
@@ -0,0 +1,547 @@
+        ReleaseDC(hwnd, hdc);
+    }
+    return hbm;
+}
+
+static VOID
+IntMakeGhostImage(HBITMAP hbm)
+{
+    BITMAP bm;
+    DWORD i, *pdw;
+
+    GetObject(hbm, sizeof(bm), &bm);
+
+    if (bm.bmBitsPixel != 32 || !bm.bmBits)
+    {
+        ERR("bm.bmBitsPixel == %d, bm.bmBits == %p\n",
+            bm.bmBitsPixel, bm.bmBits);
+        return;
+    }
+
+    pdw = bm.bmBits;
+    for (i = 0; i < bm.bmWidth * bm.bmHeight; ++i)
+    {
+        *pdw = *pdw | 0x00C0C0C0;   // bitwise-OR with ARGB #C0C0C0
+        ++pdw;
+    }
+}
+
+/****************************************************************************/
+
+typedef struct GHOST_DATA
+{
+    HWND hwndTarget;
+    HBITMAP hbm32bpp;
+} GHOST_DATA;
+
+static GHOST_DATA *
+Ghost_GetData(HWND hwnd)
+{
+    return (GHOST_DATA *)GetWindowLongPtrW(hwnd, GWLP_USERDATA);
+}
+
+static HWND
+Ghost_GetTarget(HWND hwnd)
+{
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+    if (!pData)
+        return NULL;
+    return pData->hwndTarget;
+}
+
+static BOOL
+Ghost_OnCreate(HWND hwnd, CREATESTRUCTW *lpcs)
+{
+    HBITMAP hbm32bpp;
+    HWND hwndTarget, hwndPrev;
+    GHOST_DATA *pData;
+    RECT rc;
+    DWORD style, exstyle;
+    WCHAR szNotRespondingW[64];
+    LPWSTR pszTextW = NULL;
+    INT cchTextW;
+    PWND pWnd = ValidateHwnd(hwnd);
+    if (pWnd)
+    {
+        if (!pWnd->fnid)
+        {
+            NtUserSetWindowFNID(hwnd, FNID_GHOST);
+        }
+        else if (pWnd->fnid != FNID_GHOST)
+        {
+             ERR("Wrong window class for Ghost! fnId 0x%x\n", pWnd->fnid);
+             return 0;
+        }
+    }
+
+    // get the target
+    hwndTarget = (HWND)lpcs->lpCreateParams;
+    if (!hwndTarget || !IsWindowVisible(hwndTarget) || IsIconic(hwndTarget) ||
+        GetParent(hwndTarget))
+    {
+        return FALSE;
+    }
+
+    // create user data
+    pData = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(GHOST_DATA));
+    if (!pData)
+        return FALSE;
+
+    // get window image
+    GetWindowRect(hwndTarget, &rc);
+    hbm32bpp = IntGetWindowBitmap(hwndTarget, rc.right - rc.left, rc.bottom - rc.top);
+    if (!hbm32bpp)
+    {
+        HeapFree(GetProcessHeap(), 0, pData);
+        return FALSE;
+    }
+    // make a ghost image
+    IntMakeGhostImage(hbm32bpp);
+
+    // set user data and prop
+    pData->hwndTarget = hwndTarget;
+    pData->hbm32bpp = hbm32bpp;
+    SetWindowLongPtrW(hwnd, GWLP_USERDATA, (LONG_PTR)pData);
+    SetPropW(hwndTarget, L"GhostProp", hwnd);
+
+    // get style
+    style = GetWindowLongPtrW(hwndTarget, GWL_STYLE);
+    exstyle = GetWindowLongPtrW(hwndTarget, GWL_EXSTYLE);
+
+    // get text
+    cchTextW = GetWindowTextLengthW(hwndTarget) + ARRAYSIZE(szNotRespondingW) + 1;
+    pszTextW = HeapAlloc(GetProcessHeap(), 0, cchTextW * sizeof(WCHAR));
+    if (!pszTextW)
+    {
+        DeleteObject(hbm32bpp);
+        HeapFree(GetProcessHeap(), 0, pData);
+        return FALSE;
+    }
+    GetWindowTextW(hwndTarget, pszTextW, cchTextW);
```

Comment:
I think that this would send a message to the target and will hang if the target stoped processing messages. You can avoid using messages if you use InternalGetWindowText instead.

Reply:
OK. Thanks.